Add v3 client for rest api

brentdodell opened the pull request:
@clintmiller & @tpett, can yall look over this? This is the first I've used the V3 API, and I had to do some things differently for it.

@tpett, in case you hadn't heard, we are having to use both the V2 & V3 APIs because neither has everything that we want.

brentdodell committed:
Add V3::Client

brentdodell committed:
Update V3 client to make it easier to use with api

- The V3 API expects params in 3 different places (path, query, and
body). This allows us to take any params, in a method, but have a
definition in that method that defines what params are allowed where,
and put them in the right place.
- Also, in the V3 API, some params are supposed to be nested inside an
object name param. If we pass an object name, this will handle the
nesting for us.
- See next 2 commits for usage examples
- I need to look back over the V2 docs...this may be something we need
to incorporate there as well.

brentdodell committed:
Add V3::Pipeline client

brentdodell committed:
Add V3::DealStage client

clintmiller commented on lib/active_campaign/v3/client.rb:
```diff
@@ -0,0 +1,60 @@
+require 'httparty'
```
Why `HTTParty` instead of just sticking with the other `HTTP` gem we were using?

brentdodell replied:
There was something I ran into yesterday that I was having a hard time getting HTTP.rb to do and their docs don't seem very thorough.

It may have had to do with being able to change the `base_uri` at any point through options passed to the request, which would help for those endpoints we discussed yesterday that were different from all the others, but I can't remember for sure if that was the only thing.

Are you opposed to HTTParty? From what I've seen, it seems to have a broader user base, which makes it easier when trying to figure out how to make it do something.

clintmiller replied:
Opposed? No. Just curious about rationale. Does that mean we'll use it for the V2 bits as well?

brentdodell replied:
Already are.

brentdodell replied:
...as of #8...not merged yet.

clintmiller reviewed (approved):
Generally ok with this....did have a question about using `HTTParty` vs the `HTTP` gem we had been using previously.

brentdodell committed:
Merge branch 'master' into add-v3-client-for-rest-api

* master:
  Update Code Climate reporter ID
  Update README badges to point to new repo

brentdodell committed:
Fix selecting string params

brentdodell committed:
Fix deal_stage_move

- Forgot to pass object name

brentdodell committed:
Add V3::DealTaskType client

brentdodell committed:
Add V3::DealTask client

brentdodell committed:
Add V3::Deal client

brentdodell committed:
Add V3::Connection client

brentdodell committed:
Add V3 ECommerceCustomer client

brentdodell committed:
Handle configuration with dry-configurable

brentdodell committed:
Fix name of DeepDataIntegrations clients

brentdodell committed:
Add EComOrder entity

- Add ActiveCampaign::V3::Entities::DeepDataIntegrations::EComOrder
- I really hate having to do this for 1 or 2 endpoints, but they nest
some objects in the json for create & update of E-Commerce Orders, and
this is the easiest way I've found to verify the required keys of those
nested objects.

brentdodell committed:
Add id property to EComOrder

- And make certain properties not required if id is present (it's an
update, not create)

brentdodell committed:
Add V3 ECommerceOrder client

brentdodell commented:
@clintmiller & @tpett, I believe all the functionality should be here now if yall would like to review.

There's still some cleanup to do as far as how V2 & V3 have diverged. I'd like to get V2 implemented as similarly as possible to V3 for ease of maintenance.

I also need to add more tests.

I'm more confident in the functionality of V3 compared to V2. Some of that may be bugs in their API (which I've tried to document), and some may be me not using their API correctly.

brentdodell committed:
Use ActiveCampaign::Settings as defaults

- In V3::Client initialize
- Using ActiveCampaign::Settings in a Rails initialize didn't work when
setting these outside of initialization. It would have just pulled the
environment variable, but you couldn't override it.

brentdodell committed:
Add usage instructions to README
